src/frontend/metrics/DeviceDiagnosticMetrics.ts

```diff
@@ -0,0 +1,198 @@
+import * as Sentry from '@sentry/react-native';
+import {
+  AppState,
+  Dimensions,
+  PixelRatio,
+  Platform,
+  type AppStateStatus,
+} from 'react-native';
+import {getRandomBytes} from 'expo-crypto';
+import {uint8ArrayToHex} from 'uint8array-extras';
+import * as NetInfo from '@react-native-community/netinfo';
+import * as Device from 'expo-device';
+import {getMonthlyHash} from './getMonthlyHash';
+import {sendMetricsData} from './sendMetricsData';
+import {getMetricsRequestInfo} from './getMetricsRequestInfo';
+import {storage} from '../hooks/persistedState/createPersistedState';
+import {
+  isDateValid,
+  isSameUtcMonthAndYear,
+  beginningOfNextMonthUtc,
+} from '../lib/date';
+import {setLongTimeout, clearLongTimeout} from '../lib/longTimeout';
+
+const STORAGE_KEY = 'DeviceDiagnosticMetricsLastSentAt';
+
+type DeviceDiagnosticMetricsData = {
+  type: 'device diagnostics v1';
+  monthlyDeviceHash: string;
+  brand?: string;
+  deviceType?: string;
+  isEmulator?: true;
+  manufacturer?: string;
+  model?: string;
+  os: string;
+  screen: {width: number; height: number; pixelRatio: number};
+  supportedCpuArchitectures?: string[];
+  totalMemory?: number;
+};
+
+function deviceTypeToString(deviceType: null | Device.DeviceType): string {
+  switch (deviceType) {
+    case Device.DeviceType.PHONE:
+      return 'PHONE';
+    case Device.DeviceType.TABLET:
+      return 'TABLET';
+    case Device.DeviceType.DESKTOP:
+      return 'DESKTOP';
+    case Device.DeviceType.TV:
+      return 'TV';
+    case null:
+    case Device.DeviceType.UNKNOWN:
+    default:
+      return 'UNKNOWN';
+  }
+}
+
+function getMetricsDeviceId(): string {
+  const result = storage.getString('MetricsDeviceId');
```

Comment:
I think this should be read/set at initialization, and kept in memory. Storage is fast, but it still has a cost, and there is no need to re-read this every time.

Reply:
I'd rather read it every time, especially given that this is called infrequently. Keeping the two states in sync is error-prone; for example, if someone sets the key outside this class, the in-memory value will be wrong.

I don't feel strongly, but I'm going to leave things as is. Let me know if you want to profile this or change it.